lib/auth/predicate/partial/solver.go

```diff
@@ -0,0 +1,406 @@
+func (s *Solver) PartialSolveForAll(predicate string, resolveIdentifier Resolver, querying string, to Type, timeout time.Duration) ([]z3.Value, error) {
+	outCh := make(chan []z3.Value, 1)
+	errCh := make(chan error, 1)
+
+	// Run the solver in a separate goroutine to allow the caller to handle timeouts.
+	go func() {
+		defer close(outCh)
+		defer close(errCh)
+
+		out, err := s.partialSolveForAllImpl(predicate, resolveIdentifier, querying, to)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		outCh <- out
+	}()
+
+	select {
+	case out := <-outCh:
+		return out, nil
+	case err := <-errCh:
+		return nil, err
+	case <-time.After(timeout):
+		s.def.Interrupt()
+		// wait for the other goroutine to cleanup
+		<-outCh
+		return nil, trace.LimitExceeded("timeout")
+	}
+}
+
+// partialSolveForAllImpl is the implementation of PartialSolveForAll that runs logic directly.
+func (s *Solver) partialSolveForAllImpl(predicate string, resolveIdentifier Resolver, querying string, to Type) ([]z3.Value, error) {
+	// parse the predicate expression into a Go AST
+	ast, err := parser.ParseExpr(predicate)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a context struct which is used to share various information obtained during lowering
+	ctx := &ctx{s.def, s.solver, make(map[string]z3.Value), z3.Sort{}, resolveIdentifier}
+
+	// reset the state of the Z3 solver on exit so that it can be reused
+	defer ctx.solver.Reset()
+
+	// create a Z3 identifier for the unknown value
+	switch to {
+	case TypeInt:
+		ctx.unkTy = s.def.IntSort()
+	case TypeBool:
+		ctx.unkTy = s.def.BoolSort()
+	case TypeString:
+		ctx.unkTy = s.def.StringSort()
+	default:
+		return nil, trace.BadParameter("unsupported output type %v", to)
+	}
+
+	// lower the Go AST into a Z3 AST
+	cond, err := lower(ctx, ast)
+	if err != nil {
+		return nil, err
+	}
+
+	// assert that the expression must evaluate to true
+	boolCond, ok := cond.(z3.Bool)
+	if !ok {
+		return nil, trace.BadParameter("predicate must evaluate to a boolean value")
+	}
+
+	ctx.solver.Assert(boolCond)
+	var out []z3.Value
+
+	defer func() {
+		recover()
+	}()
+
+	// retrieve all possible values for the unknown identifier
+	for {
+		// solve the model
+		sat, err := ctx.solver.Check()
+		if err != nil {
+			return nil, err
+		}
+
+		// if the model is unsatisfiable, we are done
+		if !sat {
+			return out, nil
+		}
+
+		// retrieve the model containing the output value
+		model := ctx.solver.Model()
+
+		// retrieve the value of the unknown identifier, this will error if the identifier is not found within the expression
+		val, ok := ctx.idents[querying]
+		if !ok {
+			return nil, trace.NotFound("identifier %v not found", querying)
+		}
+
+		// retrieve the value of the unknown identifier from the model
+		last := model.Eval(val, true)
+		out = append(out, last)
+
+		// assert that the value of the next possible is not equal to the current value, this allows us to retrieve all possible values
```

Comment:
what are we going to do if there are infinite possible solutions? e.g. if we allow regex matching on the login or a prefix match or something. I am not sure how to solve this but wondering if we have put any thought into whether we will prevent this or how we will handle it

Reply:
Hi, good call. To address this, I've added the option to limit the amount of solutions. That's a good assumption under almost all real world conditions anyway because we cannot display an infinite amount of possibilities.

Reply:
I don't want to block this as we've decided to go ahead with it for now, but I think we will need to solve it at some point. If you have a prefix or regex check, just setting a max number of solutions is not feasible, there are way too many options. Currently we do need to generate a complete list of possible logins and embed them in the ssh certificate and this cannot be arbitrarily large. I think we'd need some constraint on possible expressions to narrow down logins to a small finite amount, or some other expression that must return a complete list of possible logins for all nodes, or to hack our ssh server to remove this requirement.